feat(php8 attributes): building on lexer PR

cseufert opened the pull request:
Builds on #724 lexer functionality. 

PHP 8 Attribute Features:
- [x] Class
- [x] Methods
- [x] Method Params
- [x] Class Variables
- [x] Inline Functions fn()..
- [x] Functions
- [x] Anonymous Classes
- [x] Class Constants
- [x] Interfaces

cseufert commented:
Hey @ichiriac or @czosel can you guys take a look, I think it can be at least merged to the PHP 8 branch, and I think willbring the Parser up to being able to parse PHP 8.0

czosel commented:
@cseufert CI is failing because the code seems to be formatted with the current prettier version, while we're still on `v2.0.5` in this repo. I'd prefer to wait with updating prettier until the `php8` branch has landed. Can you run `npm run fix`, which should revert the formatting changes? I just tried to do it myself, but I don't have push permissions on your fork :wink:

czosel commented:
Support for constants in anonymous classes still seems to be missing:
```php
$object = new #[ObjectAttribute] class () { /* … */ };
```
fails with
```
SyntaxError: Parse Error : syntax error, unexpected '#[' (T_ATTRIBUTE) on line 2
```

czosel commented on src/parser/statement.js:
```diff
@@ -150,9 +154,17 @@ module.exports = {
+      case this.tok.T_FUNCTION: {
+        const result = this.read_function(false, false);
+        result.attrGroups = attrs;
+        attrs = [];
```
Is the reset of `attrs` really necessary?

cseufert replied:
I believe I have removed all cases where it is not strictly necessary.

czosel commented on src/lexer/attribute.js:
```diff
@@ -0,0 +1,59 @@
+      return this.consume_NUM();
+    }
+
+    throw new Error("what now? " + JSON.stringify(ch));
```
The error message could probably be improved. In `scripting.js`, we have for example:
```php
    throw new Error(
      'Bad terminal sequence "' +
        ch +
        '" at line ' +
        this.yylineno +
        " (offset " +
        this.offset +
        ")"
    );
```

cseufert replied:
Has been updated as suggested.

czosel commented on src/lexer/attribute.js:
```diff
@@ -0,0 +1,59 @@
+      this.unput(1);
+      return null;
+    }
+    // process.stderr.write("Token:" + JSON.stringify(ch) + "\n");
```
Not needed

cseufert replied:
For reasons that I dont understand, if I dont unput and return null, then all the Attribute names include leading spaces & sometimes newlines, eg:
```diff
              Attribute {
                "args": Array [],
                "kind": "attribute",
-               "name": "Two",
+               "name": "
+       Two",
              },
```

czosel commented:
@cseufert all in all this looks great already! :+1: If you'd prefer to add anonymous class support in a separate PR, let me know. The only thing this still needs to be merged in the `php8` is the formatting change and the cleanup of the few nitpicks I added above.

czosel commented on src/parser/statement.js:
```diff
@@ -150,9 +154,17 @@ module.exports = {
   read_inner_statement: function () {
+    let attrs = [];
+    if (this.token === this.tok.T_ATTRIBUTE) {
+      attrs = this.read_attr_list();
```
This line doesn't seem to be covered by the tests according to #749 (see https://github.com/glayzzle/php-parser/pull/749). Does that mean attributes in inner statements are already handled somewhere else?

cseufert replied:
It seems you can declare class methods inside inner_statements, im just not sure how to write a test to test such a case.

cseufert commented:
> @cseufert CI is failing because the code seems to be formatted with the current prettier version, while we're still on `v2.0.5` in this repo. I'd prefer to wait with updating prettier until the `php8` branch has landed. Can you run `npm run fix`, which should revert the formatting changes? I just tried to do it myself, but I don't have push permissions on your fork

I am strugling with this, yarn install did install 2.3.5, however now npm is installing it and looking at the lock file, its 2.2.3 as well.
https://github.com/glayzzle/php-parser/blob/14718704b761c462dc980a05c9eed636f5391ede/package-lock.json#L3412

cseufert commented:
@czosel I think we need to merge the master branch into the PHP8 branch, as there has been a bunch of stuff done on the main master branch that this branch has, but the upstream PHP8 branch is missing.

czosel commented:
You're right, it seems that prettier has been updated on `master` in the meantime:
```
    "prettier": {
      "version": "2.2.1",
      "resolved": "https://registry.npmjs.org/prettier/-/prettier-2.2.1.tgz",
      "integrity": "sha512-PqyhM2yCjg/oKkFPtTGUojv7gnZAoG80ttl45O6x2Ug/rMJw4wcc9k6aaf2hibP7BGVCCM33gZoGjyvt9mm16Q==",
      "dev": true
    },
```
I'd prefer to keep formatting changes out of this PR though anyway - otherwise this PR will show up in `git blame` all over the place :wink: 

On this branch, the lockfile specifies
```
    "prettier": {
      "version": "2.0.5",
      "resolved": "https://registry.npmjs.org/prettier/-/prettier-2.0.5.tgz",
      "integrity": "sha512-7PtVymN48hGcO4fGjybyBSIWDsLU4H4XlvOHfq91pz9kkGlonzwTfYkaIEwiRg/dAJF9YlbsduBAgtYLi+8cFg==",
      "dev": true
    },
```
and for me installation works as expected (using `npm` instad of `yarn`):
```sh
$ npm install
$ npm ls prettier
php-parser@3.0.2 /home/christian/dev/php-parser
└── prettier@2.0.5 
```

czosel commented:
> @czosel I think we need to merge the master branch into the PHP8 branch, as there has been a bunch of stuff done on the main master branch that this branch has, but the upstream PHP8 branch is missing.

Agreed - but I'd say let's do that in a separate PR after merging this, so that we get a clean history. Ok?

cseufert commented:
@czosel I have added a few more places that you can put attributes(interfaces, anon classes, class constants), I think I have them all now. I have cleaned up the prettier issues.

czosel commented:
@cseufert Awesome! I'll try to take another look tonight (CET).

czosel reviewed (commented):
This looks great :+1: 
Just a few minor comments, then we can merge!

cseufert commented on src/parser/expr.js:
```diff
@@ -691,7 +691,7 @@ module.exports = {
     do {
       this.expect(this.tok.T_ATTRIBUTE);
-      const attrGr = this.node("attrgroup")();
+      const attrGr = this.node("attrgroup")([]);
```
Yes, but due to the way the comments are applied with partial application, it appears you cant skip arguments, or the docs stuff gets sent as the attrib parameter, and the location gets put in the docs.

cseufert commented on src/parser/expr.js:
```diff
@@ -684,19 +684,19 @@ module.exports = {
-      this.next();
-      attrGr.attrs.push(this.read_attribute());
-      while (this.token === ",") {
+    if (this.token === this.tok.T_ATTRIBUTE)
```
Can we use eslint to enforce this, see #757 

I have fixed it in this case.

alexander-akait commented:
/cc @czosel @cseufert friendly ping, what is the blocker to merge?

czosel commented:
Nothing from my side! Sorry about the delay.
